encrypted_search/search.py

```diff
@@ -0,0 +1,124 @@
+from typing import Dict, List, Set, Tuple, cast
+
+from .models.location import Location
+from .types import (
+    Bucket,
+    FetchedFiles,
+    HierarchicalTokens,
+    KeywordWiseResults,
+    Level,
+    Locations,
+    LookupTable,
+)
+from .utils.normalizer import normalize_hierarchical
+
+
+class EncryptedSearch:
+    """Representation of the scope in which encrypted searches can be executed.
+
+        A search is divided into three steps:
+            1. Lookup: Find relevant locations in all lookup tables in the scope of this search.
+            2. Fetch: Get relevant files from content repositories. Must be completed by external client.
+            3. Locate: Collect and shortlist doc ids within fetched files.
+
+        Args:
+            lookup_tables: Tuple of lookup tables that define the scope of the search
+
+        Examples:
+            >>> index1 = EncryptedIndex([{"event": "data here"}])
+            >>> index2 = EncryptedIndex([{"other": "index"}])
+            >>> lookup_tables = (index1.lookup_table, index2.lookup_table)
+            >>> search = EncryptedSearch(lookup_tables)
+            >>> lookup_results = search.lookup("search query here")
+            >>> fetched_files = {uri: your_fetch_method(uri) for uri in lookup_results}
+            >>> event_ids = search.locate(fetched_files)
+        """
+
+    __locations: Locations
+    __lookup_tables: Tuple[LookupTable]
+    __tokens: HierarchicalTokens
+
+    def __init__(self, lookup_tables: Tuple[LookupTable, ...]):
+        self.__lookup_tables = lookup_tables
+
+    def lookup(self, query: str) -> Set[str]:
+        """Finds relevant locations in all lookup tables and returns significant MXC URIs.
+
+        Args:
+            query: Search query string
+
+        Returns:
+            Set of MXC URIs pointing to files that contain the relevant locations.
+        """
+
+        self.__locations = {}
+        locations: Dict[str, List[Location]] = {}
+        mxc_uris = set()
+        self.__tokens = normalize_hierarchical(query)
+
+        for main_token in self.__tokens:
+            # Get sub-tokens
+            sub_tokens = self.__tokens[main_token]
```

Comment:
I'm not sure I understand why we are searching with the sub-tokens.

I understand why we need them when we build the index.  Suppose the user only remembers part of what they want to find, like they only remember the domain for an email address.  So a search for "example" should return "bob@example.com".

But if the user remembers the full token, eg "bob@example.com", then why do we need to also search for "example"?

Reply:
Fixed in d9ab7ee.